### ISSUE TITLE
Internal Links are not working on merged PDF

### ISSUE THREAD

**Issue by aarkue** [aarkue/pdf-tools-web#1]
When merging multiple PDFs that contain references or links to different PDF locations, those no longer work in the merged PDF.

We already explore and collect all named destinations. We might somehow need to re-add them back to the merged PDF file.
Moreover, we will need to investigate further how references/links are implemented in PDFs.

**Comment by stefnotch** [comment]
Could this be related to https://github.com/Hopding/pdf-lib/issues/341 ? I did notice that you're using a better maintained fork of pdf-lib

**Comment by aarkue** [comment]
Yeah, the issue you linked is the relevant upstream issue :+1:. A simple API that allows manipulating Named Destinations in PDFs will likely not come soon to https://github.com/Hopding/pdf-lib or https://github.com/cantoo-scribe/pdf-lib.

I think the best way forward would be to implement this functionality ourselves, based on the lower-level APIs of pdf-lib. Initially, I started some work on this but did not get very far.
There are some functions related to the PDF Outlines Entries (like `getNamedDestinations`) already implemented, which can probably be used for this as well.

**Comment by stefnotch** [comment]
Sounds good then. I'm guessing that that is the only major pdf-lib issue?

From what I remember, there are other issues like https://github.com/Hopding/pdf-lib/issues/169 and https://github.com/Hopding/pdf-lib/issues/140 which do have workarounds.

If I want to help with implementing the named destinations API ourselves, where would I get started? I sadly know very little about the PDF file format.

**Comment by aarkue** [comment]
Great, yes, I think otherwise we can work pretty well with what pdf-lib provides!

But good that you mention the `deletePage` issue, I think the implementation in https://github.com/aarkue/pdf-tools-web/blob/main/components/ReorderPages.tsx could be improved to actually remove content correctly from the PDF file.

Regarding this issue, I think the best way to get started would be creating/gathering some sample PDF documents with internal links (Probably easy to do with a simple LaTeX PDF with a `\ref` or `\cite` ). Next, we would try to find where `pdf-lib` exposes this link information (likely named destinations + something else). After that, it would be great to try to create a PDF file using `pdf-lib` which contains internal links.

So basically:
1. [ ] Gather sample PDF files with internal links
2. [ ] Extract link information using pdf-lib
3. [ ] Create a PDF file containing internal links using pdf-lib

Finally, then combining all of this to actually preserve internal links in merged/reordered PDFs should not be too hard, hopefully.  :smile:

**Comment by stefnotch** [comment]
For what it is worth, Mozilla's pdf.js has a extremely general [`extractPages`](https://mozilla.github.io/pdf.js/api/draft/module-pdfjsLib-PDFDocumentProxy.html#extractPages) function which can merge PDFs, remove pages, etc. And it does preserve links as far as I can tell.

Here's an example of its usage
https://github.com/mozilla/pdf.js/blob/b13ec1fc3c1ddee9f987f395b69195d9ad4c1ac9/test/unit/api_spec.js#L5650

Also, the Firefox PDF viewer can now natively merge PDFs, delete pages, etc.
